### ISSUE TITLE
MemPool-backed targets

### ISSUE THREAD

**Issue by jpsamaroo** [mrufsvold/Waluigi.jl#16]
Dagger relies on MemPool.jl to provide a storage management system which provides:
- An LRU/MRU implementation for moving data between RAM and disk
- `StorageDevice` abstractions for defining arbitrary serialization/deserialization to/from some permanent storage (e.g. a CSV file on disk, a table in an SQL DB, etc.)

I have some local changes to MemPool that would enable the above abstractions to provide a lazy file source to Dagger tasks, which would be loaded on-demand once a task requires it. We could consider also implementing file sink behavior so that the output of a Dagger task or a Waluigi job would be automatically persisted through the file and be later restorable as needed.

**Comment by mrufsvold** [comment]
That would be great! I haven't yet woven `Target`s into Dagger checkpointing because I wasn't sure how to handle the following behaviors:

 - User wants to generate a Table that is larger than memory and checkpoint its chunks to Parquet. The downstream job wants to get the path to that dataset on disk, not the data itself, so that it can grab some metadata about the job without having to deal with the actual data. My understanding is that checkpoints have the expectation that `checkpoint` and `restore` are inverses of each other. 
 - The user wants to check if a checkpoint exists with a function that is less intensive than an actual call to `restore`. For example, checking if a database table already exists instead of querying for the contents of the table. Waluigi `Target`s have a separate `is_complete` function which allows custom logic for discerning if the Job should be skipped and jump straight to calling `retrieve` logic.

I am sure that these cases are supported, but I wanted to keep moving forward until I understand checkpointing a little better.

**Comment by mrufsvold** [comment]
I read over the [MemPool StorageDevice source code](https://github.com/JuliaData/MemPool.jl/blob/49bb60f6bced32950b329d1556eaccb797d0c16f/src/storage.jl#L15). I'd like to repeat my understanding to make sure I'm tracking

Imagining a Table.jl to SQL Table target implemented with the MemPool types:

* The `StorageResource` would be a database (a specific database connection?)
* The `StorageDevice` would be a table within that database
* The `StorageLeaf` would be the specific data that is now saved in a table after `move todevice!`

You'd then define (psuedo code)
```
write_to_device!(dev::SQLTable, data::Dref) = execute(dev.connection, """CREATE TABLE $(dev.table_name) ...""") # write data to table...
read_from_device(dev::SQLTable, data::Dref) = execute(dev.connection, """SELECT * FROM $(dev.table_name) ```) |> DataFrame # insert it into data somehow
delete_from_device!(dev::SQLTable, data::Dref) = execute(dev.connection, """DROP $(dev.table_name) "")
```

**Comment by jpsamaroo** [comment]
> User wants to generate a Table that is larger than memory and checkpoint its chunks to Parquet. The downstream job wants to get the path to that dataset on disk, not the data itself, so that it can grab some metadata about the job without having to deal with the actual data. My understanding is that checkpoints have the expectation that checkpoint and restore are inverses of each other.

Yeah, this is true, we fully expect `checkpoint` and `restore` to be inverses of each other. It might be possible to expose a way to directly see the associated `Chunk` in downstream tasks (we could expose a helper that pulls it from the cache), but that'd then require digging into the structure of the contained `MemPool.DRef`, which isn't currently supported (although I could also potentially be convinced that this is a useful thing generally). I assume you need this because you want to be able to wait on the file at that path to change, so that you can trigger a new invocation of the downstream job?

> The user wants to check if a checkpoint exists with a function that is less intensive than an actual call to restore. For example, checking if a database table already exists instead of querying for the contents of the table. Waluigi Targets have a separate is_complete function which allows custom logic for discerning if the Job should be skipped and jump straight to calling retrieve logic.

This is something I've also wanted for Dagger, because in a graph full of checkpoints, you want to skip all the way to the last checkpointed result and start executing from there. We could probably have a new function-typed option that checks if `restore` is possible (with the default behavior being that we just always call `restore`).

...

> The StorageResource would be a database (a specific database connection?)
> The StorageDevice would be a table within that database

This could work, but I'd probably choose to make the `StorageResource` just a URI or other kind of simple reference to the database, and the `StorageDevice` would be mostly empty - we'd then use tags from my upcoming MemPool PR to specify the table name (which would otherwise be automatically generated). We probably want to also add some state caching to `StorageDevice` or similar that could hold an open database connection.

> You'd then define (psuedo code)

Something like this :smile: The details of implementing a `StorageDevice` are kinda low-level and annoying (it's a very new API, which will probably change), so grabbing the connection handle and the table name are slightly more involved (ideally they'll become a single simple function call each), but this is approximately the idea!

**Comment by mrufsvold** [comment]
> I assume you need this because you want to be able to wait on the file at that path to change, so that you can trigger a new invocation of the downstream job?

Yes, and I often have pipelines where there are multiple steps in the chain that are handled completely at the database level. The jobs I'm scheduling locally (hopefully with Waluigi 😄 ) are just constructing SQL, kicking off a query, and then reporting that they are done, so that the next, dependent, query can be started. 

In these cases, I'd like Targets to remain flexible enough that the user could define something like `SQLQueryTarget` that runs `CREATE TABLE` on the provided query. Alternatively, another approach to Targets could be that there is a subtype of Target called `AbstractExternalProcessTarget` which does not receive data from the result of a local process, it just awaits some completion status and the work of kicking off the external process is done within the process itself. I'm shying away from the latter a bit because I like the idea that `store(::Target, data)` and `retrieve(::Target)` are the only places where side effects (loosely defined, not the Lambda Calculus version) happen.

> The details of implementing a StorageDevice are kinda low-level and annoying

I would really like to make sure that `Waluigi.AbstractTarget` is extensible by the user at a pretty low overhead so that folks can venture off and setup Azure buckets, or whatever other random service that I don't use  as targets,. I think it makes sense to roll some generally useful ones with MemPool built-in with the package, but my gut is saying that I don't want to build all the infrastructure assuming that we have a MemPool-backed target so that users don't need to learn another API. I'm not saying that's what you're suggesting, but just documenting the higher-level design goals for myself and future folks. 

All of this is making it crystal clear that I need to do some thinking on
1. What are the specific ways in `Waluigi.Target`s and `Dagger` `checkpoint`/`MemPool.StorageDevice`s overlap and diverge in functionality?
2. Are there actually two concepts living inside `Target` right now that need to be separated more than just being subtypes `Target`? Maybe it's actually a difference in `Job` type. Maybe there should be `LocalJobs` that are producing data that are needed by downstream tasks (and need a checkpoint) vs `ExternalJobs` that drive computation outside the pipeline and just need a flag for when they are complete. (As I type this, the latter feels right. But I don't like making decisions at 12:25AM).

**Comment by jpsamaroo** [comment]
> Yes, and I often have pipelines where there are multiple steps in the chain that are handled completely at the database level. The jobs I'm scheduling locally (hopefully with Waluigi smile ) are just constructing SQL, kicking off a query, and then reporting that they are done, so that the next, dependent, query can be started.

That seems reasonable. I did some thinking and believe that it's worth exposing the underlying file handle (which also includes the path) for other purposes as well, so I'll put this on my todo list to implement.

> Alternatively, another approach to Targets could be that there is a subtype of Target called AbstractExternalProcessTarget which does not receive data from the result of a local process, it just awaits some completion status and the work of kicking off the external process is done within the process itself. I'm shying away from the latter a bit because I like the idea that store(::Target, data) and retrieve(::Target) are the only places where side effects (loosely defined, not the Lambda Calculus version) happen.

Agreed, hidden side effects are no fun when developing a runtime :smile:

> I would really like to make sure that Waluigi.AbstractTarget is extensible by the user at a pretty low overhead so that folks can venture off and setup Azure buckets, or whatever other random service that I don't use as targets,. I think it makes sense to roll some generally useful ones with MemPool built-in with the package, but my gut is saying that I don't want to build all the infrastructure assuming that we have a MemPool-backed target so that users don't need to learn another API. I'm not saying that's what you're suggesting, but just documenting the higher-level design goals for myself and future folks.

Yeah, I agree that it makes sense to have it as an optional feature in case the user has a MemPool device already available.

> What are the specific ways in Waluigi.Targets and Dagger checkpoint/MemPool.StorageDevices overlap and diverge in functionality?

I also need to do some thinking on how to unify Dagger's checkpointing with MemPool storage system as well, because right now they're two separate systems that could work quite well together.
